src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs

```diff
@@ -127,176 +125,190 @@ protected internal override HttpResponseMessage Send(HttpRequestMessage request,
-                        cancellationToken.ThrowIfCancellationRequested();
+                        byte[] buffer = await _request.Content.ReadAsByteArrayAsync(_cancellationToken).ConfigureAwait(false);
+                        _cancellationToken.ThrowIfCancellationRequested();
 
-                        Task<JSObject> fetchPromise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames, optionValues, abortController, buffer);
-                        fetchResponse = await BrowserHttpInterop.CancelationHelper(fetchPromise, cancellationToken).ConfigureAwait(true);
+                        Memory<byte> bufferMemory = buffer.AsMemory();
+                        pinBuffer = bufferMemory.Pin();
+                        fetchPromise = BrowserHttpInterop.FetchBytes(_jsController, uri, _headerNames, _headerValues, _optionNames, _optionValues, pinBuffer.Value, buffer.Length);
                     }
                 }
                 else
                 {
-                    Task<JSObject> fetchPromise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames, optionValues, abortController);
-                    fetchResponse = await BrowserHttpInterop.CancelationHelper(fetchPromise, cancellationToken).ConfigureAwait(true);
+                    fetchPromise = BrowserHttpInterop.Fetch(_jsController, uri, _headerNames, _headerValues, _optionNames, _optionValues);
                 }
+                await BrowserHttpInterop.CancellationHelper(fetchPromise, _cancellationToken, _jsController).ConfigureAwait(false);
 
-                return new WasmFetchResponse(fetchResponse, abortController, abortRegistration);
+                return ConvertResponse();
             }
             catch (Exception ex)
             {
-                abortRegistration.Dispose();
-                abortController.Dispose();
+                BrowserHttpInterop.AbortRequest(_jsController);
+
+                Dispose();
                 if (ex is JSException jse)
                 {
                     throw new HttpRequestException(jse.Message, jse);
                 }
                 throw;
             }
+            finally
+            {
+                pinBuffer?.Dispose();
+                writeStream?.Dispose();
+            }
         }
 
-        private static HttpResponseMessage ConvertResponse(HttpRequestMessage request, WasmFetchResponse fetchResponse)
+        private HttpResponseMessage ConvertResponse()
         {
-#if FEATURE_WASM_THREADS
-            lock (fetchResponse.ThisLock)
+            lock (this)
             {
-#endif
-                fetchResponse.ThrowIfDisposed();
-                string? responseType = fetchResponse.FetchResponse!.GetPropertyAsString("type")!;
-                int status = fetchResponse.FetchResponse.GetPropertyAsInt32("status");
+                ThrowIfDisposed();
```

Comment:
`ThrowIfDisposed()` also takes a lock, maybe move it out of this one?

Reply:
I prefer that check to be atomic with the rest of the code below.